middleware/rate_limiter_test.go

```diff
@@ -151,83 +191,192 @@ func TestRateLimiterWithConfig(t *testing.T) {
+		BeforeFunc: func(c echo.Context) {
+			beforeRan = true
+		},
+		Store: &inMemoryStore,
+		IdentifierExtractor: func(ctx echo.Context) (string, error) {
+			return "127.0.0.1", nil
+		},
+	})
+
+	_ = mw(handler)(c)
+
+	assert.Equal(t, true, beforeRan)
+}
+
+func rateProducer(interval time.Duration, count int, f func(i int)) {
+	ticker := time.NewTicker(interval)
+	i := 0
+	quit := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				f(i)
+				if i < count {
+					i++
+					continue
+				}
+				close(quit)
+			case <-quit:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+	<-quit
 }
 
 func TestRateLimiterMemoryStore_Allow(t *testing.T) {
-	var inMemoryStore = new(RateLimiterMemoryStore)
-	inMemoryStore.rate = 1
-	inMemoryStore.burst = 3
+	var inMemoryStore = NewRateLimiterMemoryStore(1, 3, 2*time.Second)
 
 	testCases := []struct {
 		id      string
 		allowed bool
 	}{
-		{"127.0.0.1", true},
-		{"127.0.0.1", true},
-		{"127.0.0.1", true},
-		{"127.0.0.1", false},
-		{"127.0.0.1", false},
-		{"127.0.0.1", false},
-		{"127.0.0.1", false},
+		{"127.0.0.1", true},  // 0 ms
+		{"127.0.0.1", true},  // 220 ms burst #2
+		{"127.0.0.1", true},  // 440 ms burst #3
+		{"127.0.0.1", false}, // 660 ms block
+		{"127.0.0.1", false}, // 880 ms block
+		{"127.0.0.1", true},  // 1100 ms next second #1
+		{"127.0.0.2", true},  // 1320 ms allow other ip
+		{"127.0.0.1", false}, // 1540 ms no burst
+		{"127.0.0.1", false}, // 1760 ms no burst
+		{"127.0.0.1", false}, // 1980 ms no burst
+		{"127.0.0.1", true},  // 2200 ms no burst
+		{"127.0.0.1", false}, // 2420 ms no burst
+		{"127.0.0.1", false}, // 2640 ms no burst
+		{"127.0.0.1", false}, // 2860 ms no burst
+		{"127.0.0.1", true},  // 3080 ms no burst
+		{"127.0.0.1", false}, // 3300 ms no burst
 	}
 
-	for _, tc := range testCases {
+	f := func(i int) {
+		t.Logf("Running testcase #%d", i)
+		tc := testCases[i]
 		allowed := inMemoryStore.Allow(tc.id)
-
 		assert.Equal(t, tc.allowed, allowed)
 	}
+	rateProducer(220*time.Millisecond, len(testCases)-1, f)
+}
+
+func TestRateLimiterMemoryStore_cleanupStaleVisitors(t *testing.T) {
+	var inMemoryStore = NewRateLimiterMemoryStore(1, 3)
+	inMemoryStore.visitors = map[string]Visitor{
+		"A": {
+			Limiter:  rate.NewLimiter(1, 3),
+			lastSeen: time.Now(),
+		},
+		"B": {
+			Limiter:  rate.NewLimiter(1, 3),
+			lastSeen: time.Now().Add(-1 * time.Minute),
+		},
+		"C": {
+			Limiter:  rate.NewLimiter(1, 3),
+			lastSeen: time.Now().Add(-5 * time.Minute),
+		},
+		"D": {
+			Limiter:  rate.NewLimiter(1, 3),
+			lastSeen: time.Now().Add(-10 * time.Minute),
+		},
+	}
+
+	inMemoryStore.cleanupStaleVisitors()
+
+	var exists bool
+
+	_, exists = inMemoryStore.visitors["A"]
+	assert.Equal(t, true, exists)
+
+	_, exists = inMemoryStore.visitors["B"]
+	assert.Equal(t, true, exists)
+
+	_, exists = inMemoryStore.visitors["C"]
+	assert.Equal(t, false, exists)
+
+	_, exists = inMemoryStore.visitors["D"]
```

Comment:
This test didn't check that the expiration is not performed if Visitor is visited. Please perform an Allow on identifier **D** before calling **cleanupStaleVisitors()**
Maybe I'm missing something, but if you add that test you will be testing lastSeen without using time.Sleep.